pkg/whiteflag/white_flag.go

```diff
@@ -0,0 +1,247 @@
+	}
+	defer cachedTx.Release()
+	currentTx := cachedTx.GetTransaction()
+
+	if !currentTx.IsTail() {
+		return fmt.Errorf("%w: candidate tx %s is not a tail of a bundle", ErrMilestoneApprovedInvalidBundle, currentTx.GetHash())
+	}
+
+	// load up bundle to retrieve trunk and branch of the head tx
+	cachedBundle := tangle.GetCachedBundleOrNil(currentTx.GetHash())
+	if cachedBundle == nil {
+		return fmt.Errorf("%w: bundle %s of candidate tx %s doesn't exist", ErrMissingBundle, currentTx.Tx.Bundle, currentTx.GetHash())
+	}
+	defer cachedBundle.Release()
+
+	if !cachedBundle.GetBundle().IsValid() || !cachedBundle.GetBundle().ValidStrictSemantics() {
+		return fmt.Errorf("%w: bundle %s is invalid", ErrMilestoneApprovedInvalidBundle, currentTx.Tx.Bundle)
+	}
+
+	cachedBundleHeadTx := cachedBundle.GetBundle().GetHead()
+	defer cachedBundleHeadTx.Release()
+	bundleHeadTx := cachedBundleHeadTx.GetTransaction()
+	headTxTrunkHash := bundleHeadTx.GetTrunk()
+	headTxBranchHash := bundleHeadTx.GetBranch()
+
+	var cachedTrunkTx, cachedBranchTx *tangle.CachedTransaction
+	var trunkVisited, trunkConfirmed, branchVisited, branchConfirmed bool
+	var trunkConfirmedIndex, branchConfirmedIndex milestone.Index
+
+	if _, trunkVisited = visited[headTxTrunkHash]; !trunkVisited {
+		if cachedTrunkTx = tangle.GetCachedTransactionOrNil(headTxTrunkHash); cachedTrunkTx == nil {
+			return fmt.Errorf("%w: transaction %s", ErrMissingTransaction, headTxTrunkHash)
+		}
+		defer cachedTrunkTx.Release()
+		trunkConfirmed, trunkConfirmedIndex = cachedTrunkTx.GetMetadata().GetConfirmed()
+	}
+
+	if headTxTrunkHash == headTxBranchHash {
+		branchVisited = trunkVisited
+		branchConfirmedIndex = trunkConfirmedIndex
+		branchConfirmed = trunkConfirmed
+		cachedBranchTx = cachedTrunkTx
+		// no need to load branch tx
+	} else {
+		if _, branchVisited = visited[headTxBranchHash]; !branchVisited {
+			if cachedBranchTx = tangle.GetCachedTransactionOrNil(headTxBranchHash); cachedBranchTx == nil {
+				return fmt.Errorf("%w: transaction %s", ErrMissingTransaction, headTxBranchHash)
+			}
+			defer cachedBranchTx.Release()
+			branchConfirmed, branchConfirmedIndex = cachedBranchTx.GetMetadata().GetConfirmed()
+		}
+	}
+
+	// verify that head and trunk txs are indeed tails
+	if !trunkVisited && !cachedTrunkTx.GetTransaction().IsTail() {
+		return fmt.Errorf("%w: trunk tx %s of bundle head tx %s is not a tail", ErrMilestoneApprovedInvalidBundle, headTxTrunkHash, bundleHeadTx.GetHash())
+	}
+
+	if !branchVisited && !cachedBranchTx.GetTransaction().IsTail() {
+		return fmt.Errorf("%w: branch tx %s of bundle head tx %s is not a tail", ErrMilestoneApprovedInvalidBundle, headTxBranchHash, bundleHeadTx.GetHash())
+	}
+
+	// here we reached a tail of which its past cone was already visited, therefore we include its bundle
+	if (trunkVisited || tangle.SolidEntryPointsContain(headTxTrunkHash) || trunkConfirmedIndex != milestoneIndex) ||
```

Comment:
Shouldn't trunk AND branch hit one of the 3 different criterias, for the bundle to be included?
What happens if branch is a solid entry point or confirmed by an older milestone, but not trunk?
I thought we first do a deep first search via trunk...

Reply:
Another problem I see here, isn't trunkConfirmedIndex always 0 if not confirmed yet, and this way `trunkConfirmedIndex != milestoneIndex` will always be true? Same for branch.

Reply:
Maybe we can move the traversal part:
```go

	if !tangle.SolidEntryPointsContain(headTxTrunkHash) && !trunkVisited && !trunkConfirmed {
		stack.PushFront(headTxTrunkHash)
		return nil
	}

	if !tangle.SolidEntryPointsContain(headTxBranchHash) && !branchVisited && !branchConfirmed {
		stack.PushFront(headTxBranchHash)
		return nil
	}
```
before that check?

Reply:
Yeah it was supposed to be and AND, and the confirmation check was substituted with the confirmed flag.